Fix #5660: Ensure mobile-install works

BenHenning opened the pull request:
## Explanation

Fixes #5660

I've verified with local testing that ``bazel mobile-install`` works with the APK generated for each AAB target (it ends in ``_binary``, e.g. ``oppia_dev_binary``). Thus, documentation has been updated to recommend this as the main way to install the local development version of the app as ``mobile-install`` provides substantial performance benefits (see https://bazel.build/docs/mobile-install).

The ``install_*`` targets (e.g. ``install_oppia_dev``) have been removed since ``mobile-install`` should be used, instead. The existing Starlark rule & macro has been repurposed to generating a universal APK which, in turn, will be useful for specific situations where ``mobile-install`` breaks down or is not used (such as for emulator tests).

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
N/A -- this changes nothing about the build of the app, only installation pathways.

BenHenning committed:
Ensure mobile-install works.

This updates documentation after local testing, and also replaces the
install* run targets with universal APKs. Note that these APKs do NOT
work with mobile-install, and they're going to be largely ignored for
now (though they should work fine with adb install, just with worse
performance than mobile-install). They may, instead, be used for
emulator targets in the future.

BenHenning committed:
Fix incorrect detail in doc string.

BenHenning reviewed (commented):
Self-reviewed.

BenHenning commented:
PTAL @adhiamboperes.

adhiamboperes commented on oppia_android_application.bzl:
```diff
@@ -158,61 +158,57 @@ def _package_metadata_into_deployable_aab_impl(ctx):
+        "--mode=universal",
     ]
 
+    # bundletool only generats an APKs file, so the universal APK still needs to be extracted.
```
Nit:
```suggestion
    # bundletool only generate an APKs file, so the universal APK still needs to be extracted.
```

BenHenning replied:
Fixed--nice catch!

adhiamboperes commented on oppia_android_application.bzl:
```diff
@@ -158,61 +158,57 @@ def _package_metadata_into_deployable_aab_impl(ctx):
-        mnemonic = "BuildApksFromDeployAab",
-        progress_message = "Preparing AAB deploy to device",
+        arguments = generate_universal_apk_arguments,
+        mnemonic = "GenerateUniversalAPk",
```
```suggestion
        mnemonic = "GenerateUniversalAPK",
```

BenHenning replied:
Fixed.

adhiamboperes commented on oppia_android_application.bzl:
```diff
@@ -158,61 +158,57 @@ def _package_metadata_into_deployable_aab_impl(ctx):
+        inputs = [apks_file],
+        tools = [],
+        command = command,
+        mnemonic = "ExtractUniversalAPk",
```
```suggestion
        mnemonic = "ExtractUniversalAPK",
```

BenHenning replied:
Fixed.

adhiamboperes commented on oppia_android_application.bzl:
```diff
@@ -390,30 +394,30 @@ def oppia_android_application(name, config_file, proguard_generate_mapping, **kw
-    Creates a new target that can be run with 'bazel run' to install an AAB file.
+    Creates a new installable universal APK target for the provided AAB target.
+
+    This new target is installable via 'adb install'. This should generally never be used for direct
+    development installs as it will be far less performant than 'bazel mobile-install'.
```
I find this ambiguous, it is not clear to me whether the warning is against using 'adb install' or `bazel mobile-install //:oppia_prod_universal_apk`

BenHenning replied:
Ah, yeah this is just very confusing. I decided to remove the 'adb install' bit since it's not important to know, and tried to make the other parts of the documentation here a bit clearer. I did add a new 'adb install' section to cover a detail I didn't think about originally--there are cases when mobile-install can actually be detrimental to use. It's mainly useful for development since it can substantially decrease install time, but it runs different code than a real production APK so it's less good for testing production builds.

Does this look better to you?

adhiamboperes replied:
This is clear now.

adhiamboperes reviewed (changes requested):
Thanks @BenHenning! I have a few comments. PTAL.

Rd4dev commented:
@BenHenning, @adhiamboperes, 

Just wanted to request testing for compatibility with Windows regarding the `mobile-install`. Earlier, I tested it on Windows, and the app crashed immediately after installation when using mobile-install, 

>when installed with `bazel mobile-install //:oppia_dev_binary`




https://github.com/user-attachments/assets/f064887b-46b2-4f54-9fbc-4a16626dd6fb





while it worked fine when installed via `//:install_oppia_dev` (as mentioned in the CLaM chat thread). 


I’ve recently switched to Linux, and `bazel mobile-install //:oppia_dev_binary` works well here, so I’m just seeking clarity on whether the post-PR merge would still allow Windows developers to install it successfully, since the `//:install_oppia_dev` would no longer be available, or if the PR merge would address this issue.


<h1></h1>


It might be something just with my Windows setup, so thought would add it to the discussion to see if any other Windows users face the same.

adhiamboperes commented:
> @BenHenning, @adhiamboperes,
> 
> Just wanted to request testing for compatibility with Windows regarding the `mobile-install`. Earlier, I tested it on Windows, and the app crashed immediately after installation when using mobile-install,
> 
> > when installed with `bazel mobile-install //:oppia_dev_binary`
> 
>  screen-20250126-194259.mp4 
> while it worked fine when installed via `//:install_oppia_dev` (as mentioned in the CLaM chat thread).
> 
> I’ve recently switched to Linux, and `bazel mobile-install //:oppia_dev_binary` works well here, so I’m just seeking clarity on whether the post-PR merge would still allow Windows developers to install it successfully, since the `//:install_oppia_dev` would no longer be available, or if the PR merge would address this issue.
> 
> It might be something just with my Windows setup, so thought would add it to the discussion to see if any other Windows users face the same.


Thanks for flagging this @Rd4dev, I tested this `bazel mobile-install //:oppia_dev_binary` on my Mac, and recorded the stacktrace:

```
FATAL EXCEPTION: main        
 Process: org.oppia.android, PID: 9508
    java.lang.RuntimeException: Unable to instantiate application com.google.devtools.build.android.incrementaldeployment.StubApplication package org.oppia.android: java.lang.SecurityException: Writable dex file '/data/local/tmp/incrementaldeployment/org.oppia.android/dex/incremental_classes4.dex' is not allowed.
     at android.app.LoadedApk.makeApplicationInner(LoadedApk.java:1466)
     at android.app.LoadedApk.makeApplicationInner(LoadedApk.java:1395)
     at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6959)
     at android.app.ActivityThread.-$$Nest$mhandleBindApplication(Unknown Source:0)
     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2236)
     at android.os.Handler.dispatchMessage(Handler.java:106)
     at android.os.Looper.loopOnce(Looper.java:205)
     at android.os.Looper.loop(Looper.java:294)
     at android.app.ActivityThread.main(ActivityThread.java:8177)
     at java.lang.reflect.Method.invoke(Native Method)
     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
    Caused by: java.lang.SecurityException: Writable dex file '/data/local/tmp/incrementaldeployment/org.oppia.android/dex/incremental_classes4.dex' is not allowed.
```

This only occured when I used an API 34 emulator. The app works fine when built on API 29. I encountered this error during my initial setup(https://github.com/oppia/oppia-android/discussions/5643)

BenHenning commented:
Thanks @Rd4dev and @adhiamboperes, this is really helpful.

@Rd4dev per Adhiambo's comment, which SDK version were you installing to, and are you using WSL2 or building with native Windows? (I'm assuming WSL2 since native Windows builds still do not work for me).

Also, it is the intent that we maintain current support for all three platforms, so I won't submit this PR until we can confirm it working in at least the places we know it works today with the install script.

BenHenning committed:
Merge branch 'develop' into ensure-mobile-install-works

BenHenning committed:
Address reviewer comments.

BenHenning commented:
Updated to latest develop & addressed comments. PTAL @adhiamboperes.

Regarding the failure, if you were using an SDK 34 device also @Rd4dev then it seems very likely both you and @adhiamboperes ran into https://github.com/bazelbuild/bazel/issues/20456. I'm not actually sure what we should do here. I lean toward continuing with this PR even though it means SDK 34+ devices can't use mobile-install. We also have the option to use adb install for those, and we can add a caveat to the wiki accordingly until we migrate to the Starlark rules_android (which I've attempted to do a few weekends ago--it's not an easy change). WDYT?

BenHenning reviewed (commented):
Self-reviewed changes & followed up to reviewer comments.

Rd4dev commented:
@BenHenning, it seems my handset was running API 14, and after testing with a Linux setup on SDK 34 emulator, the app crashed with the same stack trace @adhiamboperes recorded. Most solutions for the API 14 behavior issue suggested making the DEX files read-only, which I believe was addressed in the bazel incremental script, though I wasn’t sure if that's possible or how to implement that directly in the codebase. Adding ADB support seems reasonable for SDK 34+ and production builds. 

Separately, since the setup now uses `mobile-install`, would it be helpful to include the optional `--start_app` flag in the wiki documentation for development?

adhiamboperes reviewed (approved):
LGTM, thanks @BenHenning.

adhiamboperes commented:
> Updated to latest develop & addressed comments. PTAL @adhiamboperes.
> 
> Regarding the failure, if you were using an SDK 34 device also @Rd4dev then it seems very likely both you and @adhiamboperes ran into [bazelbuild/bazel#20456](https://github.com/bazelbuild/bazel/issues/20456). I'm not actually sure what we should do here. I lean toward continuing with this PR even though it means SDK 34+ devices can't use mobile-install. We also have the option to use adb install for those, and we can add a caveat to the wiki accordingly until we migrate to the Starlark rules_android (which I've attempted to do a few weekends ago--it's not an easy change). WDYT?


@BenHenning, since this is not a permanent issue, and would be eliminated in future upgrades of Bazel(8+), or migrating to the Starlark rules_android, I am fine with accepting this PR with documented restrictions.

BenHenning commented:
> @BenHenning, it seems my handset was running API 14, and after testing with a Linux setup on SDK 34 emulator, the app crashed with the same stack trace @adhiamboperes recorded. Most solutions for the API 14 behavior issue suggested making the DEX files read-only, which I believe was addressed in the bazel incremental script, though I wasn’t sure if that's possible or how to implement that directly in the codebase. Adding ADB support seems reasonable for SDK 34+ and production builds.
> 
> Separately, since the setup now uses `mobile-install`, would it be helpful to include the optional `--start_app` flag in the wiki documentation for development?

I haven't tested this with `--start_app` actually, does it work correctly? If so, we could also update `.bazelrc` and have this flag be default for whenever someone uses `mobile-install`. Do you have any thoughts on this @Rd4dev or @adhiamboperes?

BenHenning committed:
Address reviewer comments.

BenHenning committed:
Merge branch 'develop' into ensure-mobile-install-works

BenHenning reviewed (commented):
PTAL @adhiamboperes and @Rd4dev at latest wording.

If you can, can you double check that both mobile-install and the `adb install` command work as expected @Rd4dev? For the former, I want to verify that `start_app` actually works and, for the latter, that the exact `adb install` command correctly installs a workable version of the app on SDK 34. If you can't, no worries--I can check this at home later as well.

Rd4dev commented:
@BenHenning, yes, I checked the commands earlier and just double-checked them now for 
- SDK 28 (using `mobile-install` with `--start_app` in the command line and also including it in the `.bazelrc`)  
- SDK 34, 35 (performing a rebuild + running `adb install bazel-bin/oppia_dev_binary.apk`). 
 
All the above mentioned runs did install a workable version of the app. 
Including `--start_app` in `.bazelrc` would be very convenient, but I’d like to confirm if it works on Mac, as there are discussions raised here - https://github.com/bazelbuild/bazel/issues/21574. 

Also, these are tested only on Linux setup.

BenHenning commented:
Ah, I actually forgot that I enabled auto-merge. Hmm. :)

Thanks so much @Rd4dev!

@adhiamboperes can you please double check the wording in the latest changes, and maybe double check `mobile-install` on OS X per @Rd4dev's latest comment? I can address any issues in a follow-up.

adhiamboperes commented:
> Ah, I actually forgot that I enabled auto-merge. Hmm. :)
> 
> Thanks so much @Rd4dev!
> 
> @adhiamboperes can you please double check the wording in the latest changes, and maybe double check `mobile-install` on OS X per @Rd4dev's latest comment? I can address any issues in a follow-up.


Double checked, everything looks good, the commands work as expected.

BenHenning commented:
Awesome, thanks much for checking again @adhiamboperes!
